portal/portal-impl/impl/src/java/org/sakaiproject/portal/charon/SkinnableCharonPortal.java

```diff
@@ -966,7 +973,36 @@ public void doLogin(HttpServletRequest req, HttpServletResponse res, Session ses
 	public void doLogout(HttpServletRequest req, HttpServletResponse res,
 			Session session, String returnPath) throws ToolException
 	{
-		
+
+		// if this is an impersonation, then reset the users old session and
+		if(impersonating()){
+			UsageSession oldSession = (UsageSession) session.getAttribute(UsageSessionService.USAGE_SESSION_KEY);
+
+			String userId = oldSession.getUserId();
+			String userEid = oldSession.getUserEid();
+			M_log.warn("returning to user: " + userId + "/" + userEid);
+			//SAK-1245
+			Vector<String> saveAttributes = new Vector<String>();
```

Comment:
Done. Surely it would be good to change it too in the original code from where this fragment of code has been copied, in the SuTool.java file

https://crucible.sakaiproject.org/browse/~br=master/Sakai.Git/tool/tool-tool/su/src/java/org/sakaiproject/tool/su/SuTool.java?r=8d974a2cc9f181a3a19fbc7f0425d44b628972e3#to197